File: runtime/ts/src/declaration_identity.test.ts
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { createHash } from 'node:crypto';
import { test } from 'node:test';
import { createValidator } from './validate.ts';
import type { TypeBinding, Validator, WireFamily } from './validate.ts';
import {
  boundDeclaration,
  canonicalDeclaration,
  declarationDigest,
  hashDeclaration,
  typeDeclaration,
  withDeclaration,
} from './declaration_identity.ts';

interface Fixture {
  name: string;
  declaration: string;
  digest: string;
  sources?: Record<string, Record<string, string>>;
}
const rows = new Map(
  (
    JSON.parse(
      readFileSync(new URL('../../../conformance/tables/declaration-digests.json', import.meta.url), 'utf8'),
    ) as { cases: Fixture[] }
  ).cases.map((row) => [row.name, row]),
);
function row(name: string): Fixture {
  const row = rows.get(name);
  assert.ok(row, name);
  return row;
}
function fixture(name: string): Validator {
  const selected = row(name),
    graph = JSON.parse(selected.declaration),
    family: WireFamily = { types: {} };
  for (const file of ['model.json', 'protocol.json', 'live.json']) {
    const source = selected.sources?.same?.[file];
    if (source) {
      const part = JSON.parse(source) as WireFamily;
      Object.assign(family.types, part.types);
      if (part.parameters) family.parameters = part.parameters;
    }
  }
  if (!graph.definitions.same) {
    graph.definitions.same = { kind: 'family', parameters: [], types: {} };
    graph.root = { ref: 'same' };
  }
  const document = canonicalDeclaration(graph);
  return withDeclaration(createValidator(family, hashDeclaration(document)), document);
}

test('canonical byte encoding and synchronous SHA-256 match every shared declaration fixture', () => {
  for (const selected of rows.values()) {
    assert.equal(canonicalDeclaration(JSON.parse(selected.declaration)), selected.declaration, selected.name);
    assert.equal(hashDeclaration(selected.declaration), selected.digest, selected.name);
  }
  for (const input of ['', 'abc', 'a'.repeat(1000000), '\u{10000}\ue000<>&\u2028\u2029'])
    assert.equal(hashDeclaration(input), createHash('sha256').update(input).digest('hex'));
  assert.equal(
    canonicalDeclaration({ '\u{10000}': 'a', '\ue000': '<>&\u2028\u2029' }),
    '{"\ue000":"\\u003c\\u003e\\u0026\\u2028\\u2029","\u{10000}":"a"}',
  );
});

test('closed identity retains ordered arguments, aliases, nested applications and same-path revisions', () => {
  const family = fixture('generic family template'),
    callable = fixture('callable template');
  const first = fixture('same named argument first revision'),
    second = fixture('same named argument second revision');
  const builtin = createValidator({ types: {} }, '');
  const integer: TypeBinding = { validate: builtin, type: 'integer' },
    text: TypeBinding = { validate: builtin, type: 'string' };
  const job1: TypeBinding = { validate: first, type: 'Job' },
    job2: TypeBinding = { validate: second, type: 'Job' };
  assert.equal(boundDeclaration(family, { A: integer, B: text }), row('bound generic family').declaration);
  assert.equal(
    boundDeclaration(family, { A: text, B: integer }),
    row('bound generic family changed arguments').declaration,
  );
  assert.equal(
    typeDeclaration({ validate: callable, type: 'Function', slots: { A: job1, B: job2 } }),
    row('same path distinct revisions in separate arguments').declaration,
  );
  assert.equal(
    typeDeclaration({ validate: callable, type: 'Function', slots: { A: job2, B: job1 } }),
    row('same path distinct revisions argument order changed').declaration,
  );
  assert.equal(typeDeclaration({ validate: callable, type: 'Alias' }), row('application baseline').declaration);
  assert.equal(
    boundDeclaration(family, { A: job1, B: integer }),
    row('bound generic family stable argument closure').declaration,
  );
  const nested: TypeBinding = { validate: callable, type: 'Function', slots: { A: integer, B: text } };
  const document = boundDeclaration(family, { A: nested, B: job1 });
  assert.ok(document.includes(row('application baseline').declaration));
  assert.equal(declarationDigest(family, { A: nested, B: job1 }), hashDeclaration(document));
  assert.notEqual(declarationDigest(family, { A: nested, B: job1 }), declarationDigest(family, { A: nested, B: job2 }));
  assert.throws(() => declarationDigest(family), /missing required binding A/);
  assert.throws(() => declarationDigest(family, { A: integer }), /missing required binding B/);
  assert.throws(
    () => typeDeclaration({ validate: createValidator({ types: { Job: { kind: 'record' } } }, ''), type: 'Job' }),
    /provenance/,
  );
  assert.throws(
    () => withDeclaration(family, row('generic family template').declaration.replace('"version":1', '"version":2')),
    /version 1/,
  );
  assert.throws(
    () => withDeclaration(createValidator({ types: {} }, 'a'.repeat(64)), row('generic family template').declaration),
    /digest/,
  );
});

test('bound family arguments retain their own slots and refuse incomplete nested interpretations', () => {
  const family = fixture('generic family template'),
    graph = JSON.parse(row('generic family template').declaration);
  graph.definitions.same.parameters = [{ name: 'F', of: 'protocol' }];
  const document = canonicalDeclaration(graph);
  const consumer = withDeclaration(
    createValidator({ types: {}, parameters: [{ name: 'F', of: 'protocol' }] }, hashDeclaration(document)),
    document,
  );
  const builtin = createValidator({ types: {} }, '');
  const integer: TypeBinding = { validate: builtin, type: 'integer' },
    text: TypeBinding = { validate: builtin, type: 'string' };
  const got = boundDeclaration(consumer, { F: { name: 'same', validate: family, slots: { A: integer, B: text } } });
  assert.ok(got.includes(row('bound generic family').declaration));
  assert.throws(
    () => boundDeclaration(consumer, { F: { name: 'same', validate: family } }),
    /missing required binding A/,
  );
});

File: runtime/go/declaration_identity_test.go
package runtime

import (
	"encoding/json"
	"os"
	"strings"
	"testing"
)

type identityFixture struct {
	Name        string                       `json:"name"`
	Declaration string                       `json:"declaration"`
	Digest      string                       `json:"digest"`
	Sources     map[string]map[string]string `json:"sources"`
}

func identityFixtures(t *testing.T) map[string]identityFixture {
	t.Helper()
	data, err := os.ReadFile("../../conformance/tables/declaration-digests.json")
	if err != nil {
		t.Fatal(err)
	}
	var table struct {
		Cases []identityFixture `json:"cases"`
	}
	if err := json.Unmarshal(data, &table); err != nil {
		t.Fatal(err)
	}
	rows := map[string]identityFixture{}
	for _, row := range table.Cases {
		rows[row.Name] = row
	}
	return rows
}

// Supply the generated provenance for the selected fixtures without invoking a
// generator from a runtime unit test. Extra source types cannot enter a selected
// argument's graph: only the fixture's selected closure is made available.
func fixtureSchema(t *testing.T, row identityFixture) *Schema {
	t.Helper()
	graph, err := readDeclaration(row.Declaration)
	if err != nil {
		t.Fatal(err)
	}
	types := map[string]any{}
	var parameters []any
	for _, file := range []string{"model.json", "protocol.json", "live.json"} {
		if source := row.Sources["same"][file]; source != "" {
			var value map[string]any
			if err := json.Unmarshal([]byte(source), &value); err != nil {
				t.Fatal(err)
			}
			if declarations, ok := value["types"].(map[string]any); ok {
				for name, def := range declarations {
					types[name] = def
				}
			}
			if list, ok := value["parameters"].([]any); ok {
				parameters = list
			}
		}
	}
	if graph.Definitions["same"] == nil {
		graph.Definitions["same"] = map[string]any{"kind": "family", "parameters": []any{}, "types": map[string]any{}}
		graph.Root = map[string]any{"ref": "same"}
	}
	document, err := graph.document()
	if err != nil {
		t.Fatal(err)
	}
	wire, _ := json.Marshal(map[string]any{"types": types, "parameters": parameters})
	return MustSchema(string(wire), declarationHash(document), nil).MustWithDeclaration(document)
}

func TestClosedDeclarationIdentitySharedBytes(t *testing.T) {
	rows := identityFixtures(t)
	for name, row := range rows {
		t.Run(name, func(t *testing.T) {
			graph, err := readDeclaration(row.Declaration)
			if err != nil {
				t.Fatal(err)
			}
			encoded, err := graph.document()
			if err != nil || encoded != row.Declaration {
				t.Fatalf("canonical bytes changed: %v", err)
			}
			if declarationHash(encoded) != row.Digest {
				t.Fatal("shared SHA-256 differs")
			}
		})
	}
}

func TestClosedDeclarationIdentityBindings(t *testing.T) {
	rows := identityFixtures(t)
	family := fixtureSchema(t, rows["generic family template"])
	callable := fixtureSchema(t, rows["callable template"])
	first := fixtureSchema(t, rows["same named argument first revision"])
	second := fixtureSchema(t, rows["same named argument second revision"])
	builtin := MustSchema(`{"types":{}}`, "", nil)
	integer := TypeBinding{Schema: builtin, Type: "integer"}
	text := TypeBinding{Schema: builtin, Type: "string"}
	job1 := TypeBinding{Schema: first, Type: "Job"}
	job2 := TypeBinding{Schema: second, Type: "Job"}
	check := func(name, got string, err error) {
		t.Helper()
		if err != nil {
			t.Fatal(err)
		}
		if got != rows[name].Declaration {
			t.Fatalf("%s differs\ngot %s\nwant %s", name, got, rows[name].Declaration)
		}
	}
	got, err := family.Bind(map[string]any{"A": integer, "B": text}, nil).BoundDeclaration()
	check("bound generic family", got, err)
	got, err = family.Bind(map[string]any{"A": text, "B": integer}, nil).BoundDeclaration()
	check("bound generic family changed arguments", got, err)
	got, err = (TypeBinding{Schema: callable.Bind(map[string]any{"A": job1, "B": job2}, nil), Type: "Function"}).Declaration()
	check("same path distinct revisions in separate arguments", got, err)
	got, err = (TypeBinding{Schema: callable.Bind(map[string]any{"A": job2, "B": job1}, nil), Type: "Function"}).Declaration()
	check("same path distinct revisions argument order changed", got, err)
	got, err = (TypeBinding{Schema: callable, Type: "Alias"}).Declaration()
	check("application baseline", got, err)
	got, err = family.Bind(map[string]any{"A": job1, "B": integer}, nil).BoundDeclaration()
	check("bound generic family stable argument closure", got, err)

	// A supplied generic value keeps its own nested type bindings.
	nested := TypeBinding{Schema: callable.Bind(map[string]any{"A": integer, "B": text}, nil), Type: "Function"}
	closed := family.Bind(map[string]any{"A": nested, "B": job1}, nil)
	outer, err := closed.BoundDeclaration()
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(outer, rows["application baseline"].Declaration) {
		t.Fatal("nested application provenance lost")
	}
	digest, err := closed.DeclarationDigest()
	if err != nil || digest != declarationHash(outer) {
		t.Fatalf("closed digest: %s %v", digest, err)
	}
	other := family.Bind(map[string]any{"A": nested, "B": job2}, nil)
	otherDigest, err := other.DeclarationDigest()
	if err != nil || digest == otherDigest {
		t.Fatal("nested same-path revision collapsed")
	}
	if _, err := family.DeclarationDigest(); err == nil || !strings.Contains(err.Error(), "missing required binding A") {
		t.Fatalf("missing slot: %v", err)
	}
	if _, err := family.Bind(map[string]any{"A": integer}, nil).DeclarationDigest(); err == nil {
		t.Fatal("partial binding acquired identity")
	}
	if _, err := (TypeBinding{Schema: MustSchema(`{"types":{"Job":{"kind":"record"}}}`, "", nil), Type: "Job"}).Declaration(); err == nil {
		t.Fatal("missing provenance acquired identity")
	}
	if _, err := family.WithDeclaration(strings.Replace(family.declaration, `"version":1`, `"version":2`, 1)); err == nil {
		t.Fatal("bad version accepted")
	}
	if _, err := MustSchema(`{"types":{}}`, strings.Repeat("a", 64), nil).WithDeclaration(rows["generic family template"].Declaration); err == nil {
		t.Fatal("wrong digest accepted")
	}
}

func TestClosedDeclarationFamilyBindings(t *testing.T) {
	rows := identityFixtures(t)
	family := fixtureSchema(t, rows["generic family template"])
	graph, _ := readDeclaration(family.declaration)
	definition := graph.Definitions["same"].(map[string]any)
	definition["parameters"] = []any{map[string]any{"name": "F", "of": "protocol"}}
	document, _ := graph.document()
	consumer := MustSchema(`{"types":{},"parameters":[{"name":"F","of":"protocol"}]}`, declarationHash(document), nil).MustWithDeclaration(document)
	bound := family.Bind(map[string]any{"A": "integer", "B": "string"}, nil)
	got, err := consumer.Bind(nil, map[string]*Schema{"F": bound}).BoundDeclaration()
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(got, rows["bound generic family"].Declaration) {
		t.Fatal("bound family argument lost its slots")
	}
	if _, err := consumer.Bind(nil, map[string]*Schema{"F": family}).BoundDeclaration(); err == nil {
		t.Fatal("unbound nested family acquired identity")
	}
}
